### PR TITLE
Improve FITS header performance

### PR DESCRIPTION
**Pull request by saimn**
This PR is to optimize the initial parsing of FITS headers, as described in #5593 (which could probably be closed then). As discussed in #5593, a long-standing problem with `io.fits` is that to access a given HDU all the previous HDU headers must be parsed. And before the lazy loading feature, all HDU headers were parsed. So this is a crucial issue for FITS files with many HDUs, and often a barrier to use them.

The main issue is that the headers must be parsed to find the structural keywords (BITPIX, NAXIS*) that are needed to know the position of the following extension. After having tried to optimize the header parsing itself (#8428), which is always good, I came to the same conclusion as @embray (https://github.com/astropy/astropy/issues/5593#issuecomment-422493941): we need a fast header parser. Which this PR implements.  

The `_BasicHeader` class is added with a "fast" parser written in Cython. It is not as fast as I would like, only twice faster as the pure Python. I think that it is possible to do better but will keep this for the future, it is already much faster that `Header`. This new class is used only during the construction of the HDU classes, where we need only a subset of standard keywords, so COMMENT, HIERARCH, and special keywords are not parsed. The cards are parsed only if needed, and cached. Typically only a few cards will be parsed, during the HDU class detection step. Then the header string is stored, and used to construct the usual `Header` object if the HDU is accessed.

One remains that remains is to know if we need to parse to full header (currently implemented), or just the first block, which would bring another interesting speedup for headers with many cards (see https://github.com/astropy/astropy/issues/5593#issuecomment-290389892 and following comments). We could maybe imagine having a keyword to trigger the full header parsing ?

To give some numbers, and testing on different files:

`astropy/io/fits/tests/data/stddata.fits`, with 2 tables and just a few
keywords, to check if there is a penalty. The PR is still faster.

```
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('astropy/io/fits/tests/data/stddata.fits', ext=1)"
```
*Before*: Mean +- std dev: 1.28 ms +- 0.03 ms
*After*: Mean +- std dev: 1.17 ms +- 0.02 ms

```
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('astropy/io/fits/tests/data/stddata.fits', ext=2)"
```
*Before*: Mean +- std dev: 2.88 ms +- 0.10 ms
*After*: Mean +- std dev: 2.30 ms +- 0.08 ms

Then on a quite extreme case, a file with 335 extensions, where each header
has ~1700 cards (with HIERARCH):

```
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('scipost_white.fits', ext=10)"
```
*Before*: Mean +- std dev: 173 ms +- 6 ms
*After*: Mean +- std dev: 26.8 ms +- 1.2 ms

```
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('scipost_white.fits', ext=60)"
```
*Before*: Mean +- std dev: 1.01 sec +- 0.01 sec
*After*: Mean +- std dev: 101 ms +- 3 ms

```
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('scipost_white.fits', ext=300)"
```
*Before*: Mean +- std dev: 5.26 sec +- 0.16 sec
*After*: Mean +- std dev: 450 ms +- 10 ms

fits_perf_scipost

And finally with a Euclid test file, with 650 extensions but a small number of keywords (15 to 20). It seems that they have more critical cases, and use `fitsio` instead of Astropy because of that.

fits_perf_euclid

ping @embray - if you have some time to have a look, it would be great to have your comments!

### CODE REVIEW

File: astropy/io/fits/card.py [hunk 1]
```diff
@@ -302,7 +300,13 @@ def value(self, value):
...
         if value is None:
             value = ''
-        oldvalue = self._value
+        try:
+            oldvalue = self.value
+        except VerifyError:
+            # probably a parsing error, falling back to the internal _value
+            # which should be None. This may happen while calling _fix_value.
+            oldvalue = self._value
+
```
Comment (saimn) [inline astropy/io/fits/card.py]: About these changes, see https://github.com/astropy/astropy/commit/b135229602acb50451c9819a722ea09a5b598132

File: astropy/io/fits/hdu/base.py [hunk 2]
```diff
@@ -133,27 +133,13 @@ class _BaseHDU(metaclass=_BaseHDUMeta):
...
-        depth-last order.  Each match_header() should identify an HDU type as
-        uniquely as possible.  Abstract types may choose to simply return False
-        or raise NotImplementedError to be skipped.
-
-        If any unexpected exceptions are raised while evaluating
-        match_header(), the type is taken to be _CorruptedHDU.
-        """
-
-        klass = _hdu_class_from_header(cls, header)
-        return super().__new__(klass)
```
Comment (saimn) [inline astropy/io/fits/hdu/base.py]: See https://github.com/astropy/astropy/commit/a9db13e8b4b4a25a2fb94489b8604150a99243b3, this was done twice.

File: astropy/io/fits/_utils.pyx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    while True:
+        # iterate on blocks
+        block = fileobj.read(BLOCK_SIZE)
+        if not block or len(block) < BLOCK_SIZE:
+            # header looks incorrect, raising exception to fall back to
+            # the full Header parsing
+            raise Exception
+
+        block_str = block.decode('ascii')
```
Comment (astrofrog) [inline astropy/io/fits/_utils.pyx]: Aren't headers supposed to be ASCII?

File: astropy/io/fits/_utils.pyx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        read_blocks.append(block_str)
+        idx = 0
+        while idx < BLOCK_SIZE:
+            # iterate on cards
+            end_idx = idx + CARD_LENGTH
+            card_image = block_str[idx:end_idx]
+            idx = end_idx
+
+            keyword = card_image[:8].strip()
+            sep_idx = card_image.find(VALUE_INDICATOR)
```
Comment (astrofrog) [inline astropy/io/fits/_utils.pyx]: Ins't the ``=`` separator always supposed to be in a specific character position? ``find`` could be slow... Ah I see what comes below. What about, to speed things up in the most common case, making the start of the 'if' statement ``if card_image[8] == '='`` and only using ``find`` if that fails?

File: astropy/io/fits/card.py [hunk 5]
```diff
@@ -283,8 +282,7 @@ def value(self):
         if self._value is not None:
             value = self._value
         elif self._valuestring is not None or self._image:
-            self._value = self._parse_value()
-            value = self._value
+            value = self._value = self._parse_value()
```
Comment (MSeifert04) [inline astropy/io/fits/card.py]: I would prefer the old style here.

File: astropy/io/fits/hdu/base.py [hunk 6]
```diff
@@ -68,6 +67,7 @@ def _hdu_class_from_header(cls, header):
                 if not (c.__module__.startswith('astropy.io.fits.') or
                         c in cls._hdu_registry):
                     continue
+                # print('try', c)
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/base.py]: Debugging leftover?

File: astropy/io/fits/_utils.pyx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+cdef Py_ssize_t CARD_LENGTH = 80
+cdef str VALUE_INDICATOR = '= '  # The standard FITS value indicator
+cdef str END_CARD = 'END' + ' ' * 77
+
+
+def parse_header(fileobj):
+    """The main method allowing to parse quickly a FITS header."""
+
+    cards = OrderedDict()
+    read_blocks = []
```
Comment (MSeifert04) [inline astropy/io/fits/_utils.pyx]: Lists can also be `cdef`'d. It won't be much faster but the `.append` call is optimized then.

File: astropy/io/fits/_utils.pyx [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    cdef int found_end = 0
+    cdef bytes block
+    cdef str header_str, block_str, card_image, keyword
+    cdef Py_ssize_t idx, end_idx, sep_idx
+
+    while True:
```
Comment (MSeifert04) [inline astropy/io/fits/_utils.pyx]: Wouldn't it make sense to use `while found_end == 0` here?

File: astropy/io/fits/_utils.pyx (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: Yes, they are supposed to. But sometimes they contain unicode chars, which will fail here, and fallback to Header, which then use [`decode_ascii`](https://github.com/astropy/astropy/blob/18515587412b6606cd3549baafb29583d91ba6a1/astropy/io/fits/util.py#L276). 
One other option, that I tried, would be to keep working on bytes here. And store the header string as bytes, which would save some memory and avoid decoding headers that are not used later. But there are some complications, like the need to check for padding bytes, and the gain was null... but I will probably try again this later.

File: astropy/io/fits/_utils.pyx (hunk 4 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: Good idea, I will try this.

File: astropy/io/fits/card.py (hunk 5 above)
Comment (saimn) [reply astropy/io/fits/card.py]: Why, multiple assignment is cool :p

File: astropy/io/fits/hdu/base.py (hunk 6 above)
Comment (saimn) [reply astropy/io/fits/hdu/base.py]: Yes, thanks.

File: astropy/io/fits/_utils.pyx (hunk 7 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: Good to know, thanks.

File: astropy/io/fits/_utils.pyx (hunk 8 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: Sure!

File: astropy/io/fits/card.py (hunk 5 above)
Comment (astrofrog) [reply astropy/io/fits/card.py]: I agree with @saimn!

Comment (pllim) [reply astropy/io/fits/card.py]: :scream:

Comment (MSeifert04) [reply astropy/io/fits/card.py]: Maybe it's just me, but it feels like an unintended misspelling of `value = self._value == self._parse_value()`. Also it reminds me of C.

Don't get me wrong I won't start a fight about it but this looks so weird to me because it's so rare that I see this in Python and since assignment is normally a statement not an expression it looks like it really shouldn't work.

File: astropy/io/fits/_utils.pyx [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+cdef Py_ssize_t CARD_LENGTH = 80
+cdef str VALUE_INDICATOR = '= '  # The standard FITS value indicator
+cdef str END_CARD = 'END' + ' ' * 77
+
+
+def parse_header(fileobj):
+    """The main method allowing to parse quickly a FITS header."""
+
+    cards = OrderedDict()
+    cdef read_blocks = []
```
Comment (MSeifert04) [inline astropy/io/fits/_utils.pyx]: I actually though `cdef list` - it's amazing that it worked 😮

**Comment by MSeifert04** [comment]
Looks mostly good to me.

What bothers me is the performance, I would've expected it to be significantly faster, maybe it's all the string operations (slicing and joining) that slow it down. Did you do any profilings?

Since the tests are passing it's probably okay. One thing we really should do is add some docs on how this internally works.

Oh and I think some of the refactoring stuff would be good to do in separate PRs. Makes the changes a lot easier to review (for example changing the imports and the indentation of arguments/`.format` calls).

File: astropy/io/fits/card.py [hunk 10]
```diff
@@ -161,7 +161,7 @@ def __init__(self, keyword=None, value=None, comment=None, **kwargs):
         self._keyword = None
         self._value = None
         self._comment = None
-
+        self._valuestring = None
```
Comment (MSeifert04) [inline astropy/io/fits/card.py]: Why was this moved?

File: astropy/io/fits/hdu/base.py [hunk 11]
```diff
@@ -868,6 +867,13 @@ class _ValidHDU(_BaseHDU, _Verify):
     def __init__(self, data=None, header=None, name=None, ver=None, **kwargs):
         super().__init__(data=data, header=header)
 
+        if (header is not None and
+                not isinstance(header, (Header, _BasicHeader))):
+            # TODO: Instead maybe try initializing a new Header object from
+            # whatever is passed in as the header--there are various types
+            # of objects that could work for this...
+            raise ValueError('header must be a Header object')
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/base.py]: Shouldn't this be a TypeError?

Comment (MSeifert04) [reply astropy/io/fits/hdu/base.py]: Okay, I see, that was just copied. In that case ignore the comment.

File: astropy/io/fits/_utils.pyx (hunk 9 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: Oups! It wasn't done on purpose but yeah it seems that cython understand this. The generated code is slightly different: 
```
+14:     cdef read_blocks = []

  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 14, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_read_blocks = __pyx_t_1;
  __pyx_t_1 = 0;
...
+29:         read_blocks.append(block_str)

    __pyx_t_9 = __Pyx_PyObject_Append(__pyx_v_read_blocks, __pyx_v_block_str); if (unlikely(__pyx_t_9 == ((int)-1))) __PYX_ERR(0, 29, __pyx_L1_error)
```
and with `cdef list`:
```
+14:     cdef list read_blocks = []

  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 14, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_read_blocks = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
...
+29:         read_blocks.append(block_str)

    __pyx_t_9 = __Pyx_PyList_Append(__pyx_v_read_blocks, __pyx_v_block_str); if (unlikely(__pyx_t_9 == ((int)-1))) __PYX_ERR(0, 29, __pyx_L1_error)
```

File: astropy/io/fits/card.py (hunk 5 above)
Comment (saimn) [reply astropy/io/fits/card.py]: I don't know, I frequently use this, typically to get a shorter variable name for a local block and avoid the lookup time on self. It's quite practical, and imo easier to understand that it references the same variable.

File: astropy/io/fits/card.py (hunk 10 above)
Comment (saimn) [reply astropy/io/fits/card.py]: Because it must be defined before `self.value` is set. Using the new header class triggered this bug (https://github.com/astropy/astropy/commit/b135229602acb50451c9819a722ea09a5b598132) in the getter/setter loop for `.value`, and the same for `.comment`.

File: astropy/io/fits/card.py (hunk 5 above)
Comment (MSeifert04) [reply astropy/io/fits/card.py]: Okay, so it's probably just a matter of getting used to it. I think we probably have to get used to it when [Python 3.8 with PEP 572 -- Assignment Expressions](https://www.python.org/dev/peps/pep-0572/) comes.

Comment (bsipocz) [reply astropy/io/fits/card.py]: wow, I should read more PEP, this looks quite interesting (though I agree that feels very unusual)

File: astropy/io/fits/_utils.pyx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    cdef str header_str, block_str, card_image, keyword
+    cdef Py_ssize_t idx, end_idx, sep_idx
+
+    while found_end == 0:
+        # iterate on blocks
+        block = fileobj.read(BLOCK_SIZE)
+        if not block or len(block) < BLOCK_SIZE:
+            # header looks incorrect, raising exception to fall back to
+            # the full Header parsing
+            raise Exception
```
Comment (MSeifert04) [inline astropy/io/fits/_utils.pyx]: Seems odd to raise a plain Exception. That way we probably swallow any exceptions that we don't expect when catching it.

I don't have a good idea though what makes sense here. Maybe `ValueError` or some custom exception like `FastHeaderParsingFailedException`? 😄

File: astropy/io/fits/_utils.pyx (hunk 4 above)
Comment (MSeifert04) [reply astropy/io/fits/_utils.pyx]: [`find`](https://docs.python.org/3/library/stdtypes.html#str.find) supports start and stop, maybe we can use these to limit the "searched region": `card_image.find(VALUE_INDICATOR, 0, 9)` (not sure if the indices are correct ... 😓)

File: astropy/io/fits/_utils.pyx (hunk 12 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: In any case there is a fallback to the Header parsing. I'm fine with using a custom exception but this will not make a difference, we need to catch any exception in the fast reader and then call Header as it have a better reporting of errors.

File: astropy/io/fits/_utils.pyx (hunk 4 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: Good to know. I already implemented something with @astrofrog 's suggestion (and forgot to push it ...), where I slice the first 8 characters before the `.find`. Not sure if it makes a big difference with using the start and stop parameters.

**Comment by saimn** [comment]
@MSeifert04 : 
> What bothers me is the performance, I would've expected it to be significantly faster, maybe it's all the string operations (slicing and joining) that slow it down. Did you do any profilings?

Yeah I would have expected a better performance as well, on the mockup I did it is much faster, and I don't know why. I tried several variants, keeping bytes, and nothing really helps. I think it is not critical for now, as even on small files it is faster, so I would prefer to move forward with this version and try to optimize later. Any idea or help to optimize will be greatly appreciated ;).

> Since the tests are passing it's probably okay. One thing we really should do is add some docs on how this internally works.

I don't think we currently have such docs? The additional complexity is rather limited compared to the mean complexity of io.fits :slightly_smiling_face:, the fast header is built for a short time, all in `_BaseHDU._readfrom_internal`, and its implementation is relatively straightforward.

> Oh and I think some of the refactoring stuff would be good to do in separate PRs. Makes the changes a lot easier to review (for example changing the imports and the indentation of arguments/`.format` calls).

I reverted the two import changes.

File: astropy/io/fits/_utils.pyx [hunk 13]
```diff
@@ -34,20 +34,20 @@ def parse_header(fileobj):
...
-                cards[keyword.upper()] = card_image
-            elif card_image == END_CARD:
-                found_end = 1
-                break
+            else:
+                sep_idx = card_image[:8].find(VALUE_INDICATOR)
+                if 0 < sep_idx < 8:
+                    keyword = card_image[:sep_idx]
+                    cards[keyword.upper()] = card_image
+                elif card_image == END_CARD:
```
Comment (astrofrog) [inline astropy/io/fits/_utils.pyx]: Maybe the END check should come before the ``find`` call? That is, first check if ``card_image[8:10]`` is the value indicator, then check if the card is the end card, then do the thing with find?

File: astropy/io/fits/_utils.pyx (hunk 4 above)
Comment (MSeifert04) [reply astropy/io/fits/_utils.pyx]: >Not sure if it makes a big difference with using the start and stop parameters.

It's probably faster because it avoids the copy. Probably not big but worth a try

File: astropy/io/fits/_utils.pyx (hunk 13 above)
Comment (MSeifert04) [reply astropy/io/fits/_utils.pyx]: Why?

**Comment by mhvk** [comment]
@saimn - overall, this surely is the right way to go, but some questions. First, how much does the Cython buy you? It makes working on it harder for others, so it should be significant, yet since this is mostly string operations, it is not obvious to me that pre-compiling is going to help much.

Second, have you checked whether basing yourself on `abc.Mapping` is costly? I have no experience with this particular class, but did find that ABCs in general are not cheap: for every creation of an instance, it is checked whether the relevant methods are present (it cannot do this at the class creation level). Perhaps worth a try to just remove it?

Though really I guess some detailed profiling is needed - which perhaps is better done as follow up.

**Comment by MSeifert04** [comment]
> overall, this surely is the right way to go, but some questions. First, how much does the Cython buy you? It makes working on it harder for others, so it should be significant, yet since this is mostly string operations, it is not obvious to me that pre-compiling is going to help much.

Interesting point. I was assuming that most of the speedup is gained because of Cython. But you're right the potentially slow parts are string operations that will be (at least likely) equally fast in pure Python.

> Though really I guess some detailed profiling is needed - which perhaps is better done as follow up.

👍 - but I am not sure if we should do it as follow-up, since this is adding a lot of complexity and "only" gives a factor 2 improvement it should be essential that we're on the right track.

**Comment by mhvk** [comment]
It's true that most typical use cases with many extensions probably involve small headers, not the 1700 card ones that your PR really shines on... Also, your tests assume a user in fact knows which extension they want, where perhaps more realistically (some of) the other extensions also get looked at.  Is there currently a cache for extensions, so that, e.g., if I look for extension 80 after having looked at 70, one starts at 70 and not back at 1?

**Comment by astrofrog** [comment]
Just a quick note that it would probably make sense to add several benchmarks to https://github.com/astropy/astropy-benchmarks for this - maybe one simple case with one HDU, and then a couple of more complex examples?

**Comment by saimn** [comment]
@astrofrog - Yes, was working on it ;), see https://github.com/astropy/astropy-benchmarks/pull/80

File: astropy/io/fits/_utils.pyx (hunk 4 above)
Comment (saimn) [reply astropy/io/fits/_utils.pyx]: It seems to make a small difference, always good to take :+1:

**Comment by saimn** [comment]
@mhvk 
> overall, this surely is the right way to go, but some questions. First, how much does the Cython buy you? It makes working on it harder for others, so it should be significant, yet since this is mostly string operations, it is not obvious to me that pre-compiling is going to help much.

The pure python implementation is indeed already very fast. The cython implementation is faster only for files with many keywords, but this is not exceptional to have at least one header with a few thousands keywords, typically the PRIMARY. So the goal with the cython version is also to be sure that the overhead of _BasicHeader is negligible. Typically if you get the PRIMARY extension, it is first parsed by the _BasicHeader, and then Header will be build using the header string parsed by the _BasicHeader. So some work is duplicated (parsing the cards from the header string), but the file is not read twice, that's the reason of storing this header string.

It is difficult to give definitive numbers, as it depends on the headers, how many cards, if they use HIERARCH, or have many COMMENT, etc.
But to give an idea, on the Euclid file presented above, with very few cards, python and cython readers are equally fast. And on the other test files, with many keywords, the difference is significant.
This is before:
```
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('scipost_white.fits', ext=60)"
Mean +- std dev: 132 ms +- 2 ms
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('scipost_white.fits', ext=300)"
Mean +- std dev: 613 ms +- 8 ms
```
and after : 
```
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('scipost_white.fits', ext=60)"
Mean +- std dev: 101 ms +- 3 ms
❯ python -m perf timeit --fast -s "from astropy.io import fits" "fits.getheader('scipost_white.fits', ext=300)"
Mean +- std dev: 450 ms +- 10 ms
```

Also from the benchmarks I just added (with many keywords), comparing the commit before cython and the head of the branch:
```
      21.4±0.03ms         21.1±0ms     0.98  io_fits.FITSHDUList.time_getheader
          198±3ms        157±0.1ms    ~0.79  io_fits.FITSHDUList.time_getheader_ext50
          316±1ms        256±0.6ms    ~0.81  io_fits.FITSHDUList.time_len
```

Also, it is only 50 lines of cython, the complexity is not huge. And the gain can be interesting. 

> Second, have you checked whether basing yourself on abc.Mapping is costly? I have no experience with this particular class, but did find that ABCs in general are not cheap: for every creation of an instance, it is checked whether the relevant methods are present (it cannot do this at the class creation level). Perhaps worth a try to just remove it?

I tested and cannot see a difference.

**Comment by saimn** [comment]
> It's true that most typical use cases with many extensions probably involve small headers, not the 1700 card ones that your PR really shines on... 

I don't know what's a typical fits file ;), there are many different conventions on what is stored in the files. For ESO files, the PRIMARY header often has hundreds or thousands of cards about the observation, conditions, instrument status, etc. It is probably not so uncommon.
I also know about people having tried to put thousands of extensions in a FITS file, and found astropy to be too slow for this. There is no reason why this use of FITS could not be possible.

> Also, your tests assume a user in fact knows which extension they want, where perhaps more realistically (some of) the other extensions also get looked at. Is there currently a cache for extensions, so that, e.g., if I look for extension 80 after having looked at 70, one starts at 70 and not back at 1?

The way I tested is just to load a given number of extensions. If you read 100 extensions you will pay the same price as if you read only the 100th one, it will just be less obvious because you probably do stuff with the extensions data or header. 
The way it works, with the lazy loading, is that you need to read the first N extensions if you want to access the N+1 one. The following extensions are not parsed, but then if you read again, it knows where to start from the N+1 one, and can continue to parse the file.

File: astropy/io/fits/card.py (hunk 5 above)
Comment (astrofrog) [reply astropy/io/fits/card.py]: Ah yes, PEP 572, the one that led to Guido stepping down from BDFL! I thought it was strange at first but I constantly find myself needing it.

**Comment by mhvk** [comment]
@saimn - thanks for the detailed answers! Good to know that the ABC stuff doesn't matter; I've on-and-off worried about it for `ShapedLikeNDArray`...

On Cython, I'm only half-convinced: it may not be a lot, but the barrier to anybody hoping to contribute to FITS is already very large (certainly, the fact that a bit of `Column` is cython has stopped me from pursuing fixes). Then again, with that argument the whole split into a base and a full header is probably a bigger issue of increased complexity. And definitely the present code has problems...

But really I haven't looked into this enough to know what would be best.

p.s. The lazy loading does sound like it already does exactly what it should...

**Comment by saimn** [comment]
So, what is the consensus about the cython extension ? 
With my biased point of view, the complexity added here is really small compared to other parts of io.fits. And even if debugging cython is more difficult, the algorithm here is straightforward so it should be easy enough to debug.

**Comment by mhvk** [comment]
@saimn - since you did the work, I think in the end you should decide! And you are certainly right that other parts are more complex. Indeed, one thing that would probably help anybody trying to understand more than the specific implementation is if in `base.py` you could add a somewhat longer comment about what is about to happen.

(The more I look at our various modules, the more I am reminded of @astrofrog's suggestion quite some time ago to use the docstrings of files to discuss the ideas behind the implementation...)

**Comment by saimn** [comment]
Ok, I added a bunch of comments about the implementation which should hopefully help to understand how things work. I think that it is good to go, and it would be good to have wider testing in the coming weeks, it just needs one approval :)

**Comment by mhvk** [comment]
@saimn - thanks, the additional comments seem very helpful. Sadly, I don't have time right now to do a full review, so will leave approval to @MSeifert04 or @astrofrog.

**Comment by bsipocz** [comment]
@saimn - could you please rebase?

**Comment by saimn** [comment]
Rebased, please have a look @MSeifert04, @astrofrog, @mhvk :)

File: astropy/io/fits/hdu/base.py (hunk 2 above)
Comment (embray) [reply astropy/io/fits/hdu/base.py]: This is still necessary if you want to instantiate an HDU instance directly (e.g. with `type.__call__`) and have it return the correct HDU class based on the header.  I don'e think that the existence of `_readfrom_internal` obviates its need because that's only used in alternative constructors for reading an HDU from data `_BaseHDU.fromstring` and `_BaseHDU.readfrom`.  That's different from calling `_BaseHDU.__new__` directly.

Comment (saimn) [reply astropy/io/fits/hdu/base.py]: Yeah I realized this later with  #8660, as some convenience methods relies on this. In #8666 I added a new class method to use the HDU class detection, i.e. `_BaseHDU._from_data(data, header)` instead of relying on `_BaseHDU.__call__`. I think it's a good compromise, it is no more possible to have an HDU class change is type based on the header which is a good thing imo, and we avoid the duplicated detection in the general case.
